### ISSUE TITLE
Bug when comparing generic/protocol types

### ISSUE THREAD

**Issue by dbrattli** [microsoft/pyright#1236]
## Environment data

-  Pylance language server 2020.11.2 (pyright 750a296b)
- OS and version: Mac OS 11.01
- Python version (& distribution if applicable, e.g. Anaconda): Python 3.9 (brew)

## Expected behaviour

There needs to be a way to compare generic types. Maybe I don't know the right way to do this, but my feeling is that there is one or more bugs here.

## Actual behaviour

Cannot compare generic types without getting type error "<" not supported for object and object.

![Screenshot 2020-12-03 at 12 54 22](https://user-images.githubusercontent.com/849479/101014845-c9973b80-3566-11eb-99cd-ef8ff5a57bd0.png)

## Logs

```
[Info  - 7:43:16 AM] Pylance language server 2020.12.1-pre.1 (pyright ea581b3f) starting
[Info  - 7:43:16 AM] Server root directory: /Users/dbrattli/.vscode/extensions/ms-python.vscode-pylance-2020.12.1-pre.1/dist
[Error - 7:43:16 AM] stubPath /Users/dbrattli/Developer/Github/Expression/typings is not a valid directory.
[Info  - 9:27:27 AM] No configuration file found.
[Info  - 9:27:27 AM] Setting pythonPath for service "Expression": "/usr/local/opt/python@3.9/bin/python3"
...
```

## Code Snippet / Additional information

```python
from typing import TypeVar

A = TypeVar("A")
B = TypeVar("B")


def compare(a: A, b: B) -> bool:
    return a < b


compare("first", "second")
```

One way to try to solve this is to make a comparable protocol. Now comparing `a < b` will type check,  but now you cannot use builtin types like `str` even if `str` do have a `__lt__` method:

![Screenshot 2020-12-03 at 12 58 22](https://user-images.githubusercontent.com/849479/101015161-432f2980-3567-11eb-8e93-20d3fc7ddc07.png)

```python
from abc import abstractmethod
from typing import Protocol, TypeVar

C = TypeVar("C", bound="Comparable")


class Comparable(Protocol):
    @abstractmethod
    def __lt__(self: C, other: C) -> bool:
        pass


AC = TypeVar("AC", bound=Comparable)
BC = TypeVar("BC", bound=Comparable)


def compare2(a: AC, b: BC) -> bool:
    return a < b


compare2("first", "second")
```

This gives error:

![Screenshot 2020-12-03 at 12 54 34](https://user-images.githubusercontent.com/849479/101015265-6d80e700-3567-11eb-9bbd-d9299c63a424.png)

**Comment by askurihin** [comment]
I think, you should declare **Comparable** as a [Generic protocol.](https://www.python.org/dev/peps/pep-0544/#generic-protocols)

This code works for me
```python
from abc import abstractmethod
from typing import Protocol, TypeVar

C = TypeVar("C")


class Comparable(Protocol[C]):
    @abstractmethod
    def __lt__(self, other: C) -> bool:
        ...


AC = TypeVar("AC", bound=Comparable)
BC = TypeVar("BC", bound=Comparable)


def compare2(a: AC, b: BC) -> bool:
    return a < b


compare2("first", "second")

```

**Comment by erictraut** [comment]
There are a few issues here.

First, when you use a TypeVar within a generic function, it should generally appear more than once. In your `compare2` function, `AC` appears only once and `BC` appears only once. That means there's nothing for the TypeVar solver to "solve" when matching these type variables.

@askurihin is on the right path. You want to use a generic protocol here. However, you can't bind to an unspecialized generic class. (Pylance/Pyright should generate an error if you try to do so. It doesn't currently, but I've filed an issue on that.)

Here's how I'd approach this problem.

```python
from abc import abstractmethod
from typing import Protocol, TypeVar

_T = TypeVar("_T")

class ComparableTo(Protocol[_T]):
    @abstractmethod
    def __lt__(self, other: _T) -> bool:
        pass

def compare2(a: ComparableTo[_T], b: _T) -> bool:
    return a < b

compare2("first", "second")
```

You'll note that Pyright generates an error on the line `return a < b`. I think that's a bug, and I'll look into it.

**Comment by dbrattli** [comment]
Thanks for the feedback. Using a generic protocol fixed my issue.

**Comment by erictraut** [comment]
I've added support for binary operator magic methods that operate on constrained TypeVars. This eliminates the error in my proposed solution above. This fix will be in the next release.

**Comment by dbrattli** [comment]
Thanks for fixing. Look forwards to test this.

**Comment by erictraut** [comment]
This is addressed in Pyright 1.1.93, which I just published. It will also be included in the next release of Pylance.
